src/hotspot/share/gc/g1/g1ConcurrentMarkThread.cpp

```diff
@@ -153,19 +188,29 @@ void G1ConcurrentMarkThread::stop_service() {
   CGC_lock->notify_all();
 }
 
-bool G1ConcurrentMarkThread::wait_for_next_cycle() {
+G1ConcurrentMarkThread::Command G1ConcurrentMarkThread::wait_for_next_cycle() {
   assert(!in_progress(), "should have been cleared");
 
   MonitorLocker ml(CGC_lock, Mutex::_no_safepoint_check_flag);
   while (!started() && !should_terminate()) {
     ml.wait();
   }
 
+  Command result;
+  if (should_terminate()) {
+    result = Terminate;
+  } else if (_state == StartMark) {
+    result = MarkCycle;
+  } else {
+    assert(_state == StartUndo, "must be");
+    result = UndoCycle;
+  }
+
```

Comment:
I would prefer if `wait_for_next_cycle()` continued returning bool and not see Terminate as a command. I'm also thinking that the command and _state is more or less the same thing, or am I missing something? If we want it could we instead have the command as a member that we set when we either start or undo/cancel the cycle?